ddtrace/internal/writer.py

```diff
@@ -487,50 +487,87 @@ def write(self, spans=None):
+                self._metrics_dist(self._write_metrics, "buffer.dropped.bytes", payload_size, tags=tags)
+            else:
+                self._metrics_dist(self._write_metrics, "buffer.accepted.traces")
+                self._metrics_dist(self._write_metrics, "buffer.accepted.spans", num_spans)
 
     def flush_queue(self, raise_exc=False):
         # type: (bool) -> None
         try:
             try:
-                n_traces = len(self._encoder)
-                encoded = self._encoder.encode()
+                with self._lock:
+                    flush_metrics = copy.deepcopy(self._write_metrics)
+                    self._write_metrics_reset()
+                    n_traces = len(self._encoder)
+                    encoded = self._encoder.encode()
+
                 if encoded is None:
                     return
             except Exception:
-                log.error("failed to encode trace with encoder %r", self._encoder, exc_info=True)
-                self._metrics_dist("encoder.dropped.traces", n_traces)
+                log.exception("failed to encode trace with encoder %r", self._encoder)
+                self._metrics_dist(flush_metrics, "encoder.dropped.traces", n_traces)
                 return
 
+            has_http_error = False
             try:
                 self._retry_upload(self._send_payload, encoded, n_traces)
-            except tenacity.RetryError as e:
-                self._metrics_dist("http.errors", tags=["type:err"])
-                self._metrics_dist("http.dropped.bytes", len(encoded))
-                self._metrics_dist("http.dropped.traces", n_traces)
+            except tenacity.RetryError as re:
                 if raise_exc:
-                    e.reraise()
-                else:
-                    log.error("failed to send traces to Datadog Agent at %s", self._agent_endpoint, exc_info=True)
+                    re.reraise()
+                has_http_error = True
+            except (HTTPException, Exception):
+                has_http_error = True
             finally:
-                if self._report_metrics and self.dogstatsd:
-                    # Note that we cannot use the batching functionality of dogstatsd because
-                    # it's not thread-safe.
-                    # https://github.com/DataDog/datadogpy/issues/439
-                    # This really isn't ideal as now we're going to do a ton of socket calls.
-                    self.dogstatsd.distribution("datadog.tracer.http.sent.bytes", len(encoded))
-                    self.dogstatsd.distribution("datadog.tracer.http.sent.traces", n_traces)
-                    for name, metric in self._metrics.items():
-                        self.dogstatsd.distribution("datadog.tracer.%s" % name, metric["count"], tags=metric["tags"])
+                self._set_http_metrics(flush_metrics, has_http_error, n_traces, len(encoded))
         finally:
-            self._set_drop_rate()
-            self._metrics_reset()
```

Comment:
metrics set in AgentWriter.write() are copied and then reset at the top of the try block. This ensures that the number of traces in the buffer is consistent with the metrics set